### ISSUE TITLE
Issues with watermark

### ISSUE THREAD

**Issue by pawanrawal** [dgraph-io/dgraph#489]
While trying to load this http://downloads.dbpedia.org/current/core/mappingbased_objects_en.ttl.bz2 faced a couple of issues. Two times the loader got stuck, the server displayed

```
[3484] Writing batch of size: 10059
[3485] Writing batch of size: 4114
2017/01/16 10:58:29 lists.go:136: Commit counters. done: 1487000 noop: 1826245
Taking snapshot for group: 1 at watermark: 9027

Writing snapshot to WAL: {Data:[9 1 0 0 0 0 0 0 0 16 1 26 15 108 111 99 97 108 104 111 115 116 58 49 50 51 52 53] Metadata:{ConfState:{Nodes:[1] XXX_unrecognized:[]} Index:9027 Term:2 XXX_unrecognized:[]} XXX_unrecognized:[]}
Current watermark 9027 <= previous snapshot 9027. Skipping.
```
in logs.

At other time the server crashed with x.Water being nil on line 336 in posting/list.go. I can't seem to reproduce it after trying it a couple of times now though.

**Comment by manishrjain** [comment]
The log is alright. It just says that it's not goint to take another snapshot; because we're caught up.

the `water` being nil is the strange case.

**Comment by squalfof** [comment]
`2017/01/19 16:38:00 lists.go:136: Commit counters. done: 35994 noop:     0
[1454] Writing batch of size: 6862
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x41de5a6]

goroutine 10473 [running]:
panic(0x45bcfe0, 0xc4200120b0)
	/usr/local/Cellar/go/1.7.4/libexec/src/runtime/panic.go:500 +0x1a1
github.com/dgraph-io/dgraph/posting.(*List).addMutation(0xc4a4451ef0, 0x4926700, 0xc46f955620, 0xc44dd51020, 0x5, 0x5, 0x415090abf3)
	/Users/xa_xx/go/src/github.com/dgraph-io/dgraph/posting/list.go:336 +0x2b6
github.com/dgraph-io/dgraph/posting.(*List).AddMutationWithIndex(0xc4a4451ef0, 0x4926700, 0xc46f955620, 0xc44dd51020, 0x0, 0x0)
	/Users/xa_xx/go/src/github.com/dgraph-io/dgraph/posting/index.go:171 +0x2c0`

Reproduced it as above, 
-- the same file, setting dgraph-mem limit to 2GB

**Comment by squalfof** [comment]
If no one's on this, can i pick this one ?

**Comment by manishrjain** [comment]
I fixed at least one part of this today. Can you pull master and double
check, if you can still reproduce this bug.

**Comment by squalfof** [comment]
still could appear, given the newest version ("Bring back client helper functions") ,  with 
posting/list.go from line 336 -> 337,
posting/index.go line 171 -> 173, diff with the last crash screenshot
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x41df816]

goroutine 35201 [running]:
panic(0x45bfd80, 0xc4200120b0)
	/usr/local/Cellar/go/1.7.4/libexec/src/runtime/panic.go:500 +0x1a1
github.com/dgraph-io/dgraph/posting.(*List).addMutation(0xc480ddfb20, 0x492a700, 0xc44dfa0990, 0xc4ab85b920, 0x4979b40, 0x5, 0x405ca20)
	/Users/xa_xx/go/src/github.com/dgraph-io/dgraph/posting/list.go:337 +0x2b6
github.com/dgraph-io/dgraph/posting.(*List).AddMutationWithIndex(0xc480ddfb20, 0x492a700, 0xc44dfa0990, 0xc4ab85b920, 0x0, 0x0)
	/Users/xa_xx/go/src/github.com/dgraph-io/dgraph/posting/index.go:173 +0x2f3
github.com/dgraph-io/dgraph/worker.runMutations(0x77a58e0, 0xc44dfa0990, 0xc4aa1ca900, 0x484, 0x520, 0x0, 0x0)
	/Users/xa_xx/go/src/github.com/dgraph-io/dgraph/worker/mutation.go:50 +0x24c
github.com/dgraph-io/dgraph/worker.(*node).processMutation(0xc4204d0000, 0x2, 0x446a, 0x0, 0xc44c4a4000, 0x6c6b, 0x6c6b, 0x0, 0x0, 0x0, ...)
	/Users/xa_xx/go/src/github.com/dgraph-io/dgraph/worker/draft.go:372 +0x13b
github.com/dgraph-io/dgraph/worker.(*node).process(0xc4204d0000, 0x2, 0x446a, 0x0, 0xc44c4a4000, 0x6c6b, 0x6c6b, 0x0, 0x0, 0x0, ...)
	/Users/xa_xx/go/src/github.com/dgraph-io/dgraph/worker/draft.go:405 +0x387
created by github.com/dgraph-io/dgraph/worker.(*node).processApplyCh
	/Users/xa_xx/go/src/github.com/dgraph-io/dgraph/worker/draft.go:444 +0x42a
```

**Comment by manishrjain** [comment]
Looking.

**Comment by manishrjain** [comment]
I was able to load up this entire data set, without any issues. Though querying the data is currently an issue: https://github.com/dgraph-io/dgraph/issues/521

**Comment by manishrjain** [comment]
Ran it twice. Works for me. Please reopen if you can reproduce this on master.

**Comment by ashwin95r** [comment]
#550

**Comment by manishrjain** [comment]
Submitted a fix for #550. Please verify the fix.

**Comment by pawanrawal** [comment]
This was verified and it works well now.
